source/common/router/shadow_writer_impl.cc

```diff
@@ -6,17 +6,25 @@
 #include "common/common/assert.h"
 #include "common/http/headers.h"
 
+#include "absl/strings/match.h"
+#include "absl/strings/str_join.h"
+
 namespace Envoy {
 namespace Router {
 
 void ShadowWriterImpl::shadow(const std::string& cluster, Http::MessagePtr&& request,
                               std::chrono::milliseconds timeout) {
+  ASSERT(request->headers().Host()->value().c_str() != "");
+
   // Switch authority to add a shadow postfix. This allows upstream logging to make a more sense.
-  // TODO PERF: Avoid copy.
-  std::string host = request->headers().Host()->value().c_str();
-  ASSERT(!host.empty());
-  host += "-shadow";
-  request->headers().Host()->value(host);
+  if (absl::StrContains(request->headers().Host()->value().c_str(), ":")) {
```

Comment:
nit: a away to rework this is to call split token then have an if on size == 2 do the StrJoin() as it is in line 23 and the else statement will the have the assert for size ==1 and use StrCat()

Reply:
Yes, this is actually my first approach on this. But had second thought hehe. Please check 7d7e613 if that's OK.